pkg/controller/volume/pvprotection/pv_protection_controller.go

```diff
@@ -0,0 +1,244 @@
+	startTime := time.Now()
+	defer func() {
+		glog.V(4).Infof("Finished processing PV %s (%v)", pvName, time.Now().Sub(startTime))
+	}()
+
+	pv, err := c.pvLister.Get(pvName)
+	if apierrs.IsNotFound(err) {
+		glog.V(4).Infof("PV %s not found, ignoring", pvName)
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	if isDeletionCandidate(pv) {
+		// PV should be deleted. Check if it's used and remove finalizer if
+		// it's not.
+		isUsed := c.isBeingUsed(pv)
+		if !isUsed {
+			return c.removeFinalizer(pv)
+		}
+	}
+
+	if needToAddFinalizer(pv) {
+		// PV is not being deleted -> it should have the finalizer. The
+		// finalizer should be added by admission plugin, this is just to add
+		// the finalizer to old PVs that were created before the admission
+		// plugin was enabled.
+		return c.addFinalizer(pv)
+	}
+	return nil
+}
+
+func (c *PVProtectionController) addFinalizer(pv *v1.PersistentVolume) error {
+	pvClone := pv.DeepCopy()
+	pvClone.ObjectMeta.Finalizers = append(pvClone.ObjectMeta.Finalizers, volumeutil.PVProtectionFinalizer)
+	_, err := c.client.CoreV1().PersistentVolumes().Update(pvClone)
+	if err != nil {
+		glog.V(3).Infof("Error adding protection finalizer to PV %s: %v", pv.Name)
+		return err
+	}
+	glog.V(3).Infof("Added protection finalizer to PV %s", pv.Name)
+	return nil
+}
+
+func (c *PVProtectionController) removeFinalizer(pv *v1.PersistentVolume) error {
+	pvClone := pv.DeepCopy()
+	pvClone.ObjectMeta.Finalizers = slice.RemoveString(pvClone.ObjectMeta.Finalizers, volumeutil.PVProtectionFinalizer, nil)
+	_, err := c.client.CoreV1().PersistentVolumes().Update(pvClone)
+	if err != nil {
+		glog.V(3).Infof("Error removing protection finalizer from PV %s: %v", pv.Name, err)
+		return err
+	}
+	glog.V(3).Infof("Removed protection finalizer from PV %s", pv.Name)
+	return nil
+}
+
+func (c *PVProtectionController) isBeingUsed(pv *v1.PersistentVolume) bool {
+	// check if PV is being bound to a PVC by its status
+	// the status will be updated by PV controller
+	if pv.Status.Phase == v1.VolumeBound {
+		// the PV is being used now
+		return true
+	}
+
+	return false
+}
+
+// pvAddedUpdated reacts to pv added/updated events
+func (c *PVProtectionController) pvAddedUpdated(obj interface{}) {
+	pv, ok := obj.(*v1.PersistentVolume)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("PV informer returned non-PV object: %#v", obj))
+		return
+	}
+	glog.V(4).Infof("Got event on PV %s", pv.Name)
+
+	if needToAddFinalizer(pv) || isDeletionCandidate(pv) {
+		c.queue.Add(pv.Name)
+	}
+}
+
+// pvcAddedDeletedUpdated reacts to PVC events
+func (c *PVProtectionController) pvcAddedDeletedUpdated(obj interface{}, deleted bool) {
+	pvc, ok := obj.(*v1.PersistentVolumeClaim)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("PVC informer returned non-PVC object: %#v", obj))
+		return
+	}
+
+	// Filter out pvc that can't help us to remove a finalizer on PV
+	if !deleted && pvc.ObjectMeta.DeletionTimestamp == nil {
```

Comment:
as you asked above, we should not delete PVs that are bound to PVCs with deletion timestamp, so checking `DeletionTimestamp` is probably useless.

Correct me if I am wrong, I think that PV controller will mark a PV as Released when a PVC is deleted so this protection controller might not need to watch PVCs at all and rely on PV controller to do the checks. Is there a case where the check above would help?

Reply:
If we are not allowed to delete PVs that are bound to PVCs with deletion timestamp,  i think you are right, PVC sruff here is useless

Reply:
If we just need to watch PV events here, maybe can consider merging it with PVC protection controller.